### PR TITLE
Provide jarmanifest/argfile approachs to shorten the command line

### LINKED ISSUE #110: CreateProcess error=206, The filename or extension is too long
/C:/Program Files/Microsoft VS Code/resources/app/out/vs/workbench/workbench.main.js:28 Failed to launch debuggee VM. Reason: java.io.IOException: Cannot run program "F:\job\2017\vscode\dev_tools\20170918\Java\jdk1.8.0_91\jre\bin\java": CreateProcess error=206, The filename or extension is too long.

### PR DESCRIPTION
**Pull request by testforstephen**
Signed-off-by: Jinbo Wang <jinbwan@microsoft.com>

Fix the bug https://github.com/Microsoft/vscode-java-debug/issues/110.

Add a new config `shortenCommandLine` in launch.json to handle long classpath cases. The possible values are `none`/`jarmanifest`/`argfile`.

Peer PR on java-debug side: https://github.com/Microsoft/java-debug/pull/253

Closes #110

### CODE REVIEW

File: package.nls.json [hunk 1]
```diff
@@ -14,6 +14,10 @@
     "java.debugger.launch.integratedTerminal.description": "VS Code integrated terminal.",
     "java.debugger.launch.externalTerminal.description": "External terminal that can be configured in user settings.",
     "java.debugger.launch.console.description": "The specified console to launch the program.",
+    "java.debugger.launch.shortenCommandLine.none": "Launch the program with the standard command line 'java [options] classname [args]'.",
+    "java.debugger.launch.shortenCommandLine.jarmanifest": "Generate the classpath parameters to a temporary classpath.jar file, and launch the program with the command line 'java -cp classpath.jar classname [args]'.",
+    "java.debugger.launch.shortenCommandLine.argfile": "Generate the classpath parameters to a temporary argument file, and launch the program with the command line 'java @argfile [args]'. This value only applies to java 9 and higher.",
```
Comment (Eskibear) [inline package.nls.json]: Java 9

File: package.nls.zh.json [hunk 2]
```diff
@@ -14,6 +14,10 @@
     "java.debugger.launch.integratedTerminal.description": "VS Code集成终端。",
     "java.debugger.launch.externalTerminal.description": "外部终端（可在用户设置中修改）。",
     "java.debugger.launch.console.description": "用于启动应用程序的控制台。",
+    "java.debugger.launch.shortenCommandLine.none": "使用标准命令行'java [options] classname [args]'启动应用程序。",
+    "java.debugger.launch.shortenCommandLine.jarmanifest": "将类路径参数生成到临时classpath.jar文件中，并使用命令行'java -cp classpath.jar classname [args]'启动应用程序。",
+    "java.debugger.launch.shortenCommandLine.argfile": "将类路径参数生成到临时argument文件中, 并使用命令行'java @argfile [args]'启动应用程序。该值仅适用于java 9及以上版本。",
```
Comment (Eskibear) [inline package.nls.zh.json]: need whitespace between Chinese char and English char

File: README.md [hunk 3]
```diff
@@ -63,6 +63,10 @@ Please also check the documentation of [Language Support for Java by Red Hat](ht
   - `internalConsole` - VS Code debug console (input stream not supported).
   - `integratedTerminal` - VS Code integrated terminal.
   - `externalTerminal` - External terminal that can be configured in user settings.
+- `shortenCommandLine` - The specified command line style to launch the program. Defaults to `none`.
+  - `none` - Launch the program with the standard command line 'java [options] classname [args]'.
```
Comment (akaroml) [inline README.md]: Add 'auto' as the default value to unblock majority of use cases.

Comment (testforstephen) [reply README.md]: A new PR https://github.com/Microsoft/java-debug/pull/254 on java-debug to help compute the launch command length.

File: README.md [hunk 4]
```diff
@@ -63,6 +63,11 @@ Please also check the documentation of [Language Support for Java by Red Hat](ht
   - `internalConsole` - VS Code debug console (input stream not supported).
   - `integratedTerminal` - VS Code integrated terminal.
   - `externalTerminal` - External terminal that can be configured in user settings.
+- `shortenCommandLine` - The specified command line style to launch the program. Defaults to `auto`.
```
Comment (yaohaizh) [inline README.md]: Should provide background for this options. In most case, user should not care about this issue. Or add a trouble shooting section for this problem.

File: package.json [hunk 5]
```diff
@@ -148,6 +148,23 @@
                 "description": "%java.debugger.launch.console.description%",
                 "default": "internalConsole"
               },
+              "shortenCommandLine": {
+                "type": "string",
+                "enum": [
+                  "auto",
```
Comment (yaohaizh) [inline package.json]: sequence: none should be put above auto

File: package.nls.json [hunk 6]
```diff
@@ -14,6 +14,11 @@
     "java.debugger.launch.integratedTerminal.description": "VS Code integrated terminal.",
     "java.debugger.launch.externalTerminal.description": "External terminal that can be configured in user settings.",
     "java.debugger.launch.console.description": "The specified console to launch the program.",
+    "java.debugger.launch.shortenCommandLine.auto": "Auto detect the command line length and determine whether to shorten the command line via the appropriate approach.",
```
Comment (yaohaizh) [inline package.nls.json]: Auto ==> "Automatically"

Should use full word in the description section. Auto standalone means car...

`the` ==> `an`

File: src/configurationProvider.ts [hunk 7]
```diff
@@ -209,6 +210,7 @@ export class JavaDebugConfigurationProvider implements vscode.DebugConfiguration
                 config.vmArgs = this.concatArgs(config.vmArgs);
             }
 
+            await populateShortenCommandLineField(config);
```
Comment (yaohaizh) [inline src/configurationProvider.ts]: Does this only apply to the launch scenario? If so, it should come together with the launch logic section.

File: src/launchCommand.ts [hunk 8]
```diff
@@ -0,0 +1,68 @@
...
+        // TODO Differentiate the limit for other platforms OSX and linux.
+        return cliLength < 32678 ? shortenApproach.none : recommendedApproach;
+    } else {
+        // In windows, the max command line length for cmd terminal is 8191 characters.
+        // TODO Differentiate the limit for other platforms OSX and linux.
+        return cliLength < 8191 ? shortenApproach.none : recommendedApproach;
+    }
+}
+
+function checkJavaVersion(javaHome: string): Promise<number> {
```
Comment (yaohaizh) [inline src/launchCommand.ts]: I think this is the third place that we need parse/check java version. Please work with @Eskibear  & @jdneo for how we can consolidate this part logic together or put in the vscode java API part.

File: src/launchCommand.ts [hunk 9]
```diff
@@ -0,0 +1,68 @@
...
+
+    return config;
+}
+
+export async function detectLaunchCommandStyle(config: vscode.DebugConfiguration): Promise<shortenApproach> {
+    const cliLength = await inferLaunchCommandLength(config);
+    const javaHome = await getJavaHome();
+    const javaVersion = await checkJavaVersion(javaHome);
+    const recommendedApproach = javaVersion <= 8 ? shortenApproach.jarmanifest : shortenApproach.argfile;
+    if (config.console === "internalConsole") {
```
Comment (yaohaizh) [inline src/launchCommand.ts]: If this command current for Windows, please add a logic such as os.platform == 'win32` to filter where this code applies

File: package.json (hunk 5 above)
Comment (testforstephen) [reply package.json]: Is there any explanation why need to sort the sequence like this?

Comment (yaohaizh) [reply package.json]: none is consistent with the original semantic

File: src/launchCommand.ts (hunk 8 above)
Comment (testforstephen) [reply src/launchCommand.ts]: agree. but prefer to do it with a new PR when the change in vscode-java is ready.

**Comment by testforstephen** [comment]
@akaroml @yaohaizh PR updated.

**Pull request merged**